### PR TITLE
Add maxlag and lags arguments to jnp.correlate and jnp.convolve

### LINKED ISSUE #37112: Custom lag arguments for np.convolve and np.correlate
Hi all,
I have an open numpy PR ([#31261](https://github.com/numpy/numpy/pull/31261)) that adds `maxlag` and `lags` keyword parameters to `np.correlate` and `np.convolve`, allowing users to compute correlation/convolution at only a subset of lags rather than the full set.  The motivation for this feature is that there are great performance improvements when the set of lags of interest is much smaller than the set of lags corresponding to any of the existing modes, as is frequently the case when analyzing long time series. There is also a new companion function `np.correlation_lags()` analogous to `scipy.signal.correlation_lags` to allow a user to generate an array of the lags that correspond to the cross-correlation/convolution they calculated.

I wanted to check in with the jax community with how big of a burden this would be on you for this feature to be added to numpy, given that you try to match the numpy API exactly.  

It seems that a possible change to jax that could match the numpy API without being too big of a lift is to add the maxlag/lags parameters to the corresponding jax functions, compute the full result as today, and slice the result appropriately before returning.  Users would get the API consistency immediately, leaving the internal performance optimization as an optional follow-on.

I am happy to put together a small PR for adding these arguments using the post-calculation slicing approach if that is helpful, but I wouldn't want to do that if it is not of interest to you.  

Best,
Honi

### PR DESCRIPTION
**Pull request by HoniSanders**
Hi all,
I have an open numpy PR ([#31261](https://github.com/numpy/numpy/pull/31261)) that adds `maxlag` and `lags` keyword parameters to `np.correlate` and `np.convolve`, allowing users to compute correlation/convolution at only a subset of lags rather than the full set.  The motivation for this feature is that there are great performance improvements when the set of lags of interest is much smaller than the set of lags corresponding to any of the existing modes, as is frequently the case when analyzing long time series. There is also a new companion function `np.correlation_lags()` analogous to `scipy.signal.correlation_lags` to allow a user to generate an array of the lags that correspond to the cross-correlation/convolution they calculated.

The numpy maintainers were concerned that changing the numpy function signature would affect jax since you try to match function signatures.  I put up an issue to get feedback with this but no one responded (https://github.com/jax-ml/jax/issues/37112).  One solution that I explained there is that a small change could allow the jax functions to accept the same arguments and give the same outputs without a major change in the underlying calculations simply by slicing appropriately.  I have implemented this in this PR for you to see what it would look like.  

Could you let me know whether this PR would be sufficient for your needs so that this functionality can be merged into numpy?

Best,
Honi

## Summary
- Adds keyword-only `maxlag` and `lags` arguments to `jnp.correlate` and
  `jnp.convolve`, plus a new `jnp.correlation_lags` companion. Tracks
  numpy/numpy#31261 and resolves jax-ml/jax#37112.
- Implementation is the simple "compute then slice" approach the issue
  proposed: when `maxlag` or `lags` is provided, pick the smallest of
  `'valid'` / `'same'` / `'full'` whose output covers the requested lags,
  run the existing `_conv` in that mode, then zero-pad (only when lags
  spill outside) and gather. Lags outside the full-overlap range yield
  zeros, matching the numpy PR's semantics.
- The no-`maxlag`/no-`lags` path is unchanged from before.

## Test plan
- [x] New `testCorrelateConvolveLags` (parametrized over op / shapes /
      lag spec / dtype)
- [x] New `testCorrelateConvolveLagsErrors` covering mode/lags conflicts,
      mutually-exclusive args, non-arithmetic-progression rejection, etc.
- [x] New `testCorrelationLags` covering mode-based and lags-based forms
- [x] Existing `testConvolutions` still passes (no regression)
- [x] `jit`, `grad`, `vmap` verified manually

## Notes for reviewers
- The public `correlate`/`convolve` lost their `@api.jit` decorator; the
  inner `_conv` is still jitted. This is to allow `lags=` to be a list /
  numpy array / jax array (those aren't hashable for `static_argnames`).
  Happy to flip to `@api.jit` on the public function and restrict `lags`
  to `range`/`slice`/`tuple` if preferred.
- Full XLA-level perf optimization (computing only requested lags via
  custom padding) is left as a follow-up.

🤖 Co-authored by [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: jax/_src/numpy/lax_numpy.py [hunk 1]
```diff
@@ -748,8 +910,70 @@ def correlate(a: ArrayLike, v: ArrayLike, mode: str = 'valid', *,
...
+
+  Examples:
+    >>> jnp.correlation_lags(5, 3, mode='full')
+    Array([-2, -1,  0,  1,  2,  3,  4], dtype=int32)
+    >>> jnp.correlation_lags(5, 3, maxlag=2)
+    Array([-2, -1,  0,  1,  2], dtype=int32)
+  """
+  lags_given = maxlag is not None or lags is not None
+  if mode is None:
+    mode = 'lags' if lags_given else 'valid'
```
[the remark this answers is not in the document]
Comment (HoniSanders) [reply jax/_src/numpy/lax_numpy.py:964]: Numpy uses default of 'valid': https://github.com/numpy/numpy/blob/27b26e9e3f95e8af3611891ccd0ff53f3169cde2/numpy/_core/numeric.py#L721

You may be thinking of scipy.signal.correlate which uses full.  https://github.com/scipy/scipy/blob/v1.17.0/scipy/signal/_signaltools.py#L104-L314

Indeed the numpy.correlation_lags implementation also uses 'valid' https://github.com/HoniSanders/numpy/blob/f67ccdcbc4de323f2880adabc9afc934780c88f5/numpy/_core/numeric.py#L879

File: jax/_src/numpy/lax_numpy.py [hunk 2]
```diff
@@ -590,9 +590,128 @@ def _conv(x: Array, y: Array, mode: str, op: str, precision: lax.PrecisionLike,
...
+  N = len(a), M = len(v).  Matches numpy's ``correlation_lags`` convention.
+  """
+  if mode == 'full':
+    return (-(M - 1), N - 1)
+  if mode == 'valid':
+    return (0, N - M) if N >= M else (N - M, 0)
+  if mode == 'same':
+    if N >= M:
+      return (-(M // 2), N - 1 - (M // 2))
+    return (N // 2 - M + 1, N // 2)
```
[the remark this answers is not in the document]
Comment (HoniSanders) [reply jax/_src/numpy/lax_numpy.py:660]: the numpy correlate docstring is clear that 'same' gives a return value of length max(M,N). 
https://github.com/numpy/numpy/blob/27b26e9e3f95e8af3611891ccd0ff53f3169cde2/numpy/_core/numeric.py#L827
